CRNS-264: Fixing issues around network recovery

vishalnarkhede opened the pull request:
# Submit a pull request

Dependent on [#626](https://github.com/GetStream/stream-chat-js/pull/626)

## CLA

- [ ] I have signed the [Stream CLA](https://docs.google.com/forms/d/e/1FAIpQLScFKsKkAJI7mhCr7K9rEIOpqIDThrWxuvxnwUq2XkHyG154vQ/viewform) (required).
- [ ] The code changes follow best practices
- [ ] Code changes are tested (add some information if not applicable)

## Description of the pull request

vishalnarkhede commented on src/components/ChannelPreview/ChannelPreviewMessenger.tsx:
```diff
@@ -197,6 +197,10 @@ export type ChannelPreviewMessengerProps<
     'channel' | 'latestMessagePreview'
   >;
 
+export const MemoizedChannelPreviewMessengerWithContext = React.memo(
```
Saves some re-renders!!

vishalnarkhede commented on examples/SampleApp/src/components/ScreenHeader.tsx:
```diff
@@ -137,6 +137,18 @@ export const ScreenHeader: React.FC<ScreenHeaderProps> = (props) => {
     }
   }, [insets.top]);
 
+  const rightContent = useMemo(() => {
+    return RightContent && 
+      <View style={styles.rightContainer}>
+        <RightContent />
+      </View>
+  }, []);
+
+  const leftContent = useMemo(() => {
+    return LeftContent && 
+        <LeftContent />
+  }, []);
+
```
Fixes the image flickering issue on sample app!!

nhannah commented on src/components/Chat/hooks/useAppStateListener.ts:
```diff
@@ -0,0 +1,59 @@
+    };
+  }, []);
+
+  const handleAppStateChange = async (nextAppState: AppStateStatus) => {
```
This will be an issue; if you get a phone call lets say, the app state will go from active -> inactive (no connection closed), then you end call and in app, app state will go from inactive -> active (new connection), now 2x connections.

vishalnarkhede replied:
This is now handled on js client level :)

nhannah replied:
we should still get the logic right tho? I.e. if we only close when going to background only open when coming from background??

vishalnarkhede replied:
If app is in inactive state for too long, websocket may still be dropped. So I think it's just better to call openConnection anyways.

nhannah replied:
Actually curious if that happens, do you know?

nhannah commented on examples/SampleApp/src/components/ScreenHeader.tsx:
```diff
@@ -137,6 +137,18 @@ export const ScreenHeader: React.FC<ScreenHeaderProps> = (props) => {
     }
   }, [insets.top]);
 
+  const rightContent = useMemo(() => {
```
Could we do these with React.memo for more consistency in the lib?

vishalnarkhede replied:
Going to handle this separately. Removed the changes from this PR!!

nhannah reviewed (changes requested):
Mostly minor comments, but one def needs changing

vishalnarkhede commented:
@nhannah Updated PR with few more changes.

- Added prop `closeConnectionOnbackground` (boolean), to be able to disable AppState listener
- Moved function-props on FlatList to separate functions, instead defining them inline.
- The things that discussed about multiple calls to client.openConnection - we have not handled it on js client level. So all good there.
- I have removed those image flickering related changes from this PR. I will handle them separately.

nhannah commented on src/components/MessageList/MessageList.tsx:
```diff
@@ -825,36 +830,57 @@ const MessageListWithContext = <
         data={messageList}
-        /** Disables the MessageList UI. Which means, message actions, reactions won't work. */
         extraData={disabled || !channel?.state.isUpToDate}
+        /** Disables the MessageList UI. Which means, message actions, reactions won't work. */
```
should this comment still be above extraData?

vishalnarkhede replied:
Done!!

nhannah commented on src/components/Chat/Chat.tsx:
```diff
@@ -162,9 +176,16 @@ const ChatWithContext = <
+
   useEffect(() => {
     if (client.setUserAgent) {
       client.setUserAgent(`stream-chat-react-native-${Platform.OS}-${version}`);
```
There will be a merge conflict with the {version} change just merged heads up.

vishalnarkhede replied:
Will be fine I think. There is no diff here on this PR. Github would have raised it by now. Will pull master anyways, when I fix the `Chat.js` comment.

vishalnarkhede replied:
Done!! Synced with master!!

nhannah commented on src/components/Channel/Channel.tsx:
```diff
@@ -639,23 +663,28 @@ const ChannelWithContext = <
 
-  const listenToChanges = () => {
+  useEffect(() => {
     // The more complex sync logic is done in Chat.js
```
.tsx can change ;) prolly old comment

vishalnarkhede replied:
Done!!

nhannah reviewed (commented):
Just had a few comments that should be easy to answer, if it's all working think were good to go though.

vishalnarkhede commented on src/components/MessageList/MessageList.tsx:
```diff
@@ -825,36 +830,57 @@ const MessageListWithContext = <
+    <HeaderComponent loadingMore={onStartReachedInProgress} />
+  );
+
+  const renderEmptyStateIndicator = () => (
```
So rendering empty state is still in flatlist, I have just separated out inline fuction. Inline functions can cause more re-renders. React documentation only gives headsup about this for ref callback function (inline vs separate function), it doesn't say any such thing about indicators. But from the discussions I can see, it's still recommended to not have inline functions in FlatList.

For loading indicator, I had to take it out of FlatList because
- when you are in network recovery state, we show loading indicator.
- after recovery, when new results are ready, scroll doesn't start at bottom of the list if loading indicator is part of FlatList (within ListEmptyComponent).

And this seemed like a confusing UX to me.

nhannah commented on src/components/Channel/Channel.tsx:
```diff
@@ -641,23 +665,28 @@ const ChannelWithContext = <
+      client.off('connection.changed', connectionChangedHandler);
+      channel?.off(handleEvent);
+    };
+  }, [channelId]);
```
```suggestion
  }, [channelId]);
  
```
